posts/2022-07-25-keyword-generics.md

```diff
@@ -0,0 +1,344 @@
+
+To give you a quick taste of what we're working on, this is how we imagine you
+may want to write a function which is generic over "asyncness" sometime in the
+future:
+
+```rust
+async<A> trait Read {
+    async<A> fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+    async<A> fn read_to_string(&mut self, buf: &mut String) -> Result<usize> { ... }
+}
+
+/// Read from a reader into a string.
+async<A> fn read_to_string(reader: &mut impl Read * A) -> std::io::Result<String> {
+    let mut string = String::new();
+    reader.read_to_string(&mut string).await?;
+    string
+}
+```
+
+This function introduces a "keyword generic" parameter into the function of `A`.
+You can think of this as a flag which indicates whether the function is being
+compiled in an async context or not. The parameter `A` is forwarded to the `impl
+Read`, making that conditional on "asyncness" as well.
+
+In the function body you can see a `.await` call. Because [the `.await` keyword
+marks cancellation sites][cancel] we unfortunately can't just infer them
+[^cancellation]. Instead we require them to be written for when the code is
+compiled in async mode, but are essentially reduced to a no-op in non-async
+mode.
+
+[cancel]: https://blog.yoshuawuyts.com/async-cancellation-1/
+[^cancellation]: No really, we can't just infer them - and it may not be as
+simple as omitting all `.await` calls either. The Async WG is working through
+the full spectrum of cancellation sites, async drop, and more. But for now we're
+working under the assumption that `.await` will remain relevant going forward.
+And even in the off chance that it isn't, fallibility has similar requirements
+at the call site as async does.
+
+We still have lots of details left to figure out, but we hope this at least
+shows the general *feel* of what we're going for. We want to make it
+significantly easier to write code which works in both sync and non-async Rust.
+
+## A peek at the past: horrors before const
+
+Rust didn't always have `const` as part of the language. A long long long long
+long time ago (2018) we had to write a regular function for runtime computations
+and associated const of generic type logic for compile-time computations. As an
+example, to add the number `1` to a constant provided to you, you had to write
+([playground]):
+
+[playground]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50e818b79b8af322ed4384d3c33e9773
+
+```rust
+trait Const<T> {
+    const VAL: T;
+}
+
+/// `42` as a "const" (type) generic:
+struct FourtyTwo;
+impl Const<i32> for FourtyTwo {
+    const VAL: i32 = 42;
+}
+
+/// `C` -> `C + 1` operation:
+struct AddOne<C: Const<i32>>(C);
+impl<C: Const<i32>> Const<i32> for AddOne<C> {
+    const VAL: i32 = C::VAL + 1;
+}
+
+AddOne::<FourtyTwo>::VAL
+```
+
+Today this is as easy as writing a `const fn`:
+
+```rust
+const fn add_one(i: i32) -> i32 {
+    i + 1
+}
+
+add_one(42)
```

Comment:
I feel like this example is missing something, since to my knowledge `42 + 1` worked for a long time already in const contexts (e.g., `const F: u32 = 42 + 1;`). The only time you might need the complex invocation is if you're using this as a const generic, but there a `const fn` isn't really going to help you either.

Maybe I'm the one missing something though :)

Reply:
you can always inline a `const fn` at its call site. After a certain complexity that is not useful though. Showing a complex example here is not very helpful to the reader either. So I chose to write a simple `add_one` function, but honestly it could be anything that doesn't require iteration or branching (as that wasn't supported before `const fn`)

Reply:
Can we mark this as resolved then?